File: apple/RecKit/Sources/RecKit/Design/BlueprintDialog.swift
import SwiftUI

/// docs/09 화면 원칙 5 ("제목 + 설명 + 최대 2개 버튼"), drawn the way the rest of the app is drawn: a
/// square-cornered node on the grid, not the platform's own alert. `alert` and `confirmationDialog`
/// are containers with their own shape, their own material and their own corner radius, and none of
/// those is in docs/09 — so this is the same surface, [Blueprint.line] border and [Radius.node]
/// corner as `StateNode`, with no shadow.
///
/// The Android shell's `BlueprintDialog` is the same component, and this mirrors it line for line:
/// the body scrolls on its own so a long one (the disconnect warnings) never pushes the answers off
/// the screen, and the answers stack when a row of them no longer fits (docs/09 유동 타이포).
///
/// There is no motion to reduce: nothing here animates.
public struct BlueprintDialog<Actions: View, Content: View>: View {
    @Environment(\.blueprint) private var blueprint
    private let title: String
    private let actions: Actions
    private let content: Content

    public init(
        title: String,
        @ViewBuilder actions: () -> Actions,
        @ViewBuilder content: () -> Content
    ) {
        self.title = title
        self.actions = actions()
        self.content = content()
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(verbatim: title)
                .font(blueprint.fonts.title)
                .foregroundStyle(blueprint.palette.text)
                // docs/09 유동 타이포: three lines is what the header may take. Every title here is
                // a short question; the one thing that can outgrow it is the user's own recording
                // name inside "Delete ‘…’?", and cutting *that* in the middle is what keeps the
                // question itself — its first words and its question mark — readable.
                .lineLimit(Self.titleLines)
                .truncationMode(.middle)
                .frame(maxWidth: .infinity, alignment: .leading)
            // The card is as tall as its question and no taller: a two-line dialog is a two-line
            // card. What scrolls when the body outgrows the screen is the page it sits on
            // ([BlueprintDialogSheet]), not a region inside it — a scroll view here would claim
            // every point the sheet offered and leave the card stretched to the bottom of it.
            VStack(alignment: .leading, spacing: Space.s) {
                content
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            // docs/09 "접근성" · 유동 타이포: the answers go across while they fit and downwards
            // when they do not — an answer clipped to a syllable makes the question unanswerable.
            // Stacked they keep their order, which puts the primary one at the bottom: last, as it
            // is last on the right, and nearest the thumb.
            ViewThatFits(in: .horizontal) {
                HStack(spacing: Space.s) {
                    Spacer(minLength: 0)
                    actions
                }
                VStack(spacing: Space.s) {
                    actions
                }
            }
        }
        .padding(Space.m)
        .frame(maxWidth: Self.maxWidth)
        .background(blueprint.palette.surface, in: RoundedRectangle(cornerRadius: Radius.node))
        .overlay {
            RoundedRectangle(cornerRadius: Radius.node)
                .strokeBorder(blueprint.palette.grid, lineWidth: blueprint.line)
        }
    }

    /// How much of a long question stays in the header.
    static var titleLines: Int { 3 }

    /// Wide enough for the disconnect warnings, narrow enough that a dialog on a Mac window still
    /// reads as something sitting *on* the screen rather than as another screen.
    static var maxWidth: CGFloat { 420 }
}

extension View {
    /// The dialog as the platform presents a modal: a sheet, over the dot grid, with the card in
    /// the middle of it. The one presentation both shells use for a question that has to be
    /// answered before anything else happens.
    public func blueprintDialog<Content: View>(
        isPresented: Binding<Bool>,
        @ViewBuilder content: @escaping () -> Content
    ) -> some View {
        sheet(isPresented: isPresented) {
            BlueprintDialogSheet { content() }
        }
    }

    /// The same, driven by the thing the dialog is about — which is also what carries the counts it
    /// has to say (how many parts, how many recordings).
    public func blueprintDialog<Item: Identifiable, Content: View>(
        item: Binding<Item?>,
        @ViewBuilder content: @escaping (Item) -> Content
    ) -> some View {
        sheet(item: item) { value in
            BlueprintDialogSheet { content(value) }
        }
    }
}

extension View {
    /// The dialog drawn *in* the screen rather than presented over it: a scrim across what the
    /// question is about, and the card on top of it.
    ///
    /// For the two places a sheet cannot go. A view may host one sheet, and a screen that already
    /// has one — the record screen's naming prompt, which is a different question with a different
    /// answer — has no second one to give. And a `LSUIElement` menu-bar app has no window at all:
    /// its popover *is* the surface, so its dialogs are drawn inside it.
    public func blueprintDialogOverlay<Content: View>(
        isPresented: Binding<Bool>,
        @ViewBuilder content: @escaping () -> Content
    ) -> some View {
        overlay {
            if isPresented.wrappedValue {
                BlueprintDialogScrim { content() }
            }
        }
    }
}

/// The dark behind a dialog drawn inside a screen: it dims what the question is about without
/// hiding it, and swallows the taps that would otherwise reach what is underneath.
///
/// docs/09 트렌드 7 keeps material for chrome, so this is a flat wash of the page colour rather than
/// a blur — there is nothing here for glass to be the chrome of.
public struct BlueprintDialogScrim<Content: View>: View {
    @Environment(\.blueprint) private var blueprint
    private let content: Content

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        ZStack {
            blueprint.palette.background.opacity(0.86)
            ScrollView {
                content.padding(Space.m).frame(maxWidth: .infinity)
            }
            .scrollBounceBehavior(.basedOnSize)
        }
        .contentShape(Rectangle())
        // The scrim is the dialog: a screen reader that could still reach the list behind it would
        // be reading a screen the question has taken over.
        .accessibilityAddTraits(.isModal)
    }
}

/// The page a [BlueprintDialog] is presented on: the dot grid, and the card centred on it.
public struct BlueprintDialogSheet<Content: View>: View {
    private let content: Content

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        ScrollView {
            content
                .padding(Space.m)
                .frame(maxWidth: .infinity)
        }
        .scrollBounceBehavior(.basedOnSize)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .dotGridBackground()
        #if os(iOS)
        .presentationDetents([.medium, .large])
        #else
        .frame(minWidth: 460, minHeight: 300)
        #endif
    }
}

/// docs/09: the body is 14–16, and what it means decides the colour.
public enum DialogTone: Sendable {
    case body
    case muted
    case danger
}

/// One line of a dialog body.
public struct BlueprintDialogText: View {
    @Environment(\.blueprint) private var blueprint
    private let text: String
    private let tone: DialogTone

    public init(_ text: String, tone: DialogTone = .body) {
        self.text = text
        self.tone = tone
    }

    public var body: some View {
        Text(verbatim: text)
            .font(tone == .body ? blueprint.fonts.body : blueprint.fonts.bodySmall)
            .foregroundStyle(ink)
            .fixedSize(horizontal: false, vertical: true)
            .frame(maxWidth: .infinity, alignment: .leading)
    }

    private var ink: Color {
        switch tone {
        case .body: return blueprint.palette.text
        case .muted: return blueprint.palette.textMuted
        case .danger: return blueprint.palette.danger
        }
    }
}

/// One of several answers, as one accessibility element: what VoiceOver reads is
/// "<label>, selected" and not an unnamed mark beside a label it cannot see (docs/09 "접근성"). The
/// row is [minTouch] tall whatever the label does.
public struct BlueprintRadioRow: View {
    private let label: String
    private let selected: Bool
    private let action: () -> Void

    public init(_ label: String, selected: Bool, action: @escaping () -> Void) {
        self.label = label
        self.selected = selected
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            OptionRow(label: label, selected: selected, filled: false)
        }
        .buttonStyle(.plain)
        .accessibilityAddTraits(selected ? [.isButton, .isSelected] : .isButton)
    }
}

/// The same row for an answer that is on or off rather than one of a set — announced as the switch
/// it is through `accessibilityRepresentation`, the way [BlueprintSwitchStyle] is.
public struct BlueprintCheckRow: View {
    private let label: String
    @Binding private var isOn: Bool

    public init(_ label: String, isOn: Binding<Bool>) {
        self.label = label
        _isOn = isOn
    }

    public var body: some View {
        Button { isOn.toggle() } label: {
            OptionRow(label: label, selected: isOn, filled: true)
        }
        .buttonStyle(.plain)
        .accessibilityRepresentation {
            Toggle(isOn: $isOn) { Text(verbatim: label) }
        }
    }
}

private struct OptionRow: View {
    @Environment(\.blueprint) private var blueprint
    let label: String
    let selected: Bool
    let filled: Bool

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: Space.s) {
            SelectionMark(selected: selected, filled: filled)
                // The mark sits on the first line of a label that wrapped, not in the middle of it.
                .alignmentGuide(.firstTextBaseline) { $0[.bottom] - 3 }
            Text(verbatim: label)
                .font(blueprint.fonts.bodySmall)
                .foregroundStyle(blueprint.palette.text)
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(.vertical, Space.s)
        .frame(minHeight: minTouch)
        .contentShape(Rectangle())
    }
}

/// docs/09 "형태": no circles and no pills, so both marks are squares on the badge radius. A radio
/// holds a smaller square inside its outline; a checkbox fills, because "on" is a state and "this
/// one of the two" is a position.
///
/// The outline is on [BlueprintPalette.line] — the token the hairline and every bordered node take
/// — so high contrast thickens it here too (docs/09 "고대비 모드"). An unchecked box is nothing
/// *but* its outline, which is what made it the one that needed it.
private struct SelectionMark: View {
    @Environment(\.blueprint) private var blueprint
    let selected: Bool
    let filled: Bool

    var body: some View {
        let ink = selected ? blueprint.palette.accent : blueprint.palette.textMuted
        RoundedRectangle(cornerRadius: Radius.badge)
            .fill(selected && filled ? ink : .clear)
            .overlay {
                if selected {
                    RoundedRectangle(cornerRadius: 1)
                        .fill(filled ? blueprint.palette.onAccent : ink)
                        .frame(width: 8, height: 8)
                }
            }
            .overlay {
                RoundedRectangle(cornerRadius: Radius.badge)
                    .strokeBorder(ink, lineWidth: blueprint.line)
            }
            .frame(width: 18, height: 18)
    }
}

/// A link inside a dialog — the Google permissions page, the consent guidance. Accent *and*
/// underlined, because a link that is only a colour is invisible to a colour-blind reader (docs/09
/// "모든 상태는 색 + 텍스트"), and [minTouch] tall because it is something you tap.
public struct BlueprintDialogLink: View {
    @Environment(\.blueprint) private var blueprint
    private let label: String
    private let action: () -> Void

    public init(_ label: String, action: @escaping () -> Void) {
        self.label = label
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            Text(verbatim: label)
                .font(blueprint.fonts.bodySmall)
                .foregroundStyle(blueprint.palette.accent)
                .underline()
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: .infinity, minHeight: minTouch, alignment: .leading)
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }
}

File: android/app/src/main/kotlin/app/recly/android/ui/component/BlueprintDialog.kt
package app.recly.android.ui.component

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.selection.toggleable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.Layout
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Constraints
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import app.recly.android.ui.theme.BlueprintColors
import app.recly.android.ui.theme.MinTouch
import app.recly.android.ui.theme.Radius
import app.recly.android.ui.theme.Space
import app.recly.android.ui.theme.blueprint

/**
 * docs/09 화면 원칙 5 ("제목 + 설명 + 최대 2개 버튼"), drawn the way the rest of the app is drawn: a
 * square-cornered node on the grid, not Material's tonal card. `AlertDialog` is a container with
 * its own shape, its own elevation tint and its own 28dp corners, and none of those are in docs/09
 * — so this is `Dialog` plus the same surface, 1dp border and 4dp radius as [StateNode].
 *
 * There is no motion to reduce: a `Dialog` window is not animated, so the system's own reduce-motion
 * setting has nothing to switch off here. Dark arrives through the palette.
 *
 * [content] scrolls on its own so a long body (the disconnect warnings) never pushes [actions] off
 * the screen; the card is capped at [dialogMaxHeight] of the window. docs/09 §유동 타이포 makes the
 * font size the user's, so the two things that grow with it are both bounded: the title keeps
 * [TITLE_LINES] in the header and spills the rest into that same scroll region, and [actions] stack
 * when a row of them would no longer fit (see [DialogActions]).
 */
@Composable
fun BlueprintDialog(
    title: String,
    onDismissRequest: () -> Unit,
    actions: @Composable () -> Unit,
    modifier: Modifier = Modifier,
    content: @Composable ColumnScope.() -> Unit,
) {
    val palette = blueprint
    val shape = RoundedCornerShape(Radius.node)
    val maxHeight = dialogMaxHeight(LocalConfiguration.current.screenHeightDp)
    // What did not fit in the header. Clipping the rest away is not on offer — a question with its
    // end cut off, or ellipsed, is not a question — so it is re-shown at the top of the body, which
    // scrolls. Empty for every title that fits, which is all of them at the default font size.
    var spill by remember(title) { mutableStateOf("") }
    Dialog(onDismissRequest = onDismissRequest) {
        Column(
            modifier = modifier
                .fillMaxWidth()
                .heightIn(max = maxHeight)
                .border(palette.line, palette.grid, shape)
                .background(palette.surface, shape)
                .padding(Space.m),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text(
                title,
                style = MaterialTheme.typography.titleLarge,
                color = palette.text,
                maxLines = TITLE_LINES,
                overflow = TextOverflow.Clip,
                onTextLayout = { layout ->
                    spill = if (layout.hasVisualOverflow) {
                        title.substring(layout.getLineEnd(layout.lineCount - 1, visibleEnd = true)).trimStart()
                    } else {
                        ""
                    }
                },
            )
            Column(
                modifier = Modifier.weight(1f, fill = false).verticalScroll(rememberScrollState()),
                verticalArrangement = Arrangement.spacedBy(Space.s),
            ) {
                if (spill.isNotEmpty()) {
                    Text(spill, style = MaterialTheme.typography.titleLarge, color = palette.text)
                }
                content()
            }
            DialogActions(actions)
        }
    }
}

/**
 * The answers, in a row while they fit and in a stack when they do not. What decides it is nearly
 * always the user's font size and not the width of the device (docs/09 §유동 타이포), and a clipped
 * answer makes the question unanswerable — so the row is measured rather than assumed.
 *
 * Stacked, each answer is full width and they keep their order, which puts the primary one at the
 * bottom: last, the same as it is last on the right, and nearest the thumb.
 */
@Composable
private fun DialogActions(actions: @Composable () -> Unit) {
    val spacing = with(LocalDensity.current) { Space.s.roundToPx() }
    Layout(content = actions, modifier = Modifier.fillMaxWidth()) { measurables, constraints ->
        val width = constraints.maxWidth
        // maxIntrinsicWidth and not a trial measure: a measurable may only be measured once, and
        // this has to know the answer before it can choose the constraints to measure with.
        val natural = measurables.map { it.maxIntrinsicWidth(Constraints.Infinity) }
        val stacked = stackActions(width, natural, spacing)
        val placeables = measurables.map { measurable ->
            measurable.measure(
                if (stacked) Constraints.fixedWidth(width) else Constraints(maxWidth = width),
            )
        }
        val gaps = spacing * (placeables.size - 1).coerceAtLeast(0)
        val height = if (stacked) {
            placeables.sumOf { it.height } + gaps
        } else {
            placeables.maxOfOrNull { it.height } ?: 0
        }
        layout(width, height) {
            if (stacked) {
                var y = 0
                placeables.forEach { placeable ->
                    placeable.placeRelative(0, y)
                    y += placeable.height + spacing
                }
            } else {
                // Aligned to the end, as [Arrangement.End] had them.
                var x = width - placeables.sumOf { it.width } - gaps
                placeables.forEach { placeable ->
                    placeable.placeRelative(x, (height - placeable.height) / 2)
                    x += placeable.width + spacing
                }
            }
        }
    }
}

/**
 * Whether the answers have to stack: they do the moment the row they would make is wider than the
 * card, gaps included.
 */
internal fun stackActions(available: Int, widths: List<Int>, spacing: Int): Boolean =
    widths.sum() + spacing * (widths.size - 1).coerceAtLeast(0) > available

/** How much of a long question stays pinned above the scroll region. */
private const val TITLE_LINES = 3

/**
 * How tall the card may get before its body scrolls: most of the window, never all of it, so the
 * dialog still reads as something sitting *on* the screen rather than as another screen.
 */
fun dialogMaxHeight(screenHeightDp: Int): Dp = (screenHeightDp * HEIGHT_FRACTION).dp

private const val HEIGHT_FRACTION = 0.8f

/** docs/09: the body is 14–16, and what it means decides the colour. */
enum class DialogTone { BODY, MUTED, DANGER }

/** One line of a dialog body. */
@Composable
fun BlueprintDialogText(text: String, modifier: Modifier = Modifier, tone: DialogTone = DialogTone.BODY) {
    val palette = blueprint
    Text(
        text,
        modifier = modifier,
        style = if (tone == DialogTone.BODY) {
            MaterialTheme.typography.bodyLarge
        } else {
            MaterialTheme.typography.bodyMedium
        },
        color = when (tone) {
            DialogTone.BODY -> palette.text
            DialogTone.MUTED -> palette.textMuted
            DialogTone.DANGER -> palette.danger
        },
    )
}

/**
 * One of several answers, as one accessibility node: the row is `selectable`, so what a screen
 * reader focuses is "<label>, radio button, selected" and not an unnamed mark beside a label
 * (docs/09 "접근성"). The row is [MinTouch] tall whatever the label does.
 */
@Composable
fun BlueprintRadioRow(
    label: String,
    selected: Boolean,
    onSelect: () -> Unit,
    modifier: Modifier = Modifier,
) {
    OptionRow(
        label = label,
        modifier = modifier.selectable(selected = selected, role = Role.RadioButton, onClick = onSelect),
        mark = { SelectionMark(selected = selected, filled = false) },
    )
}

/** The same row as [BlueprintRadioRow] for an answer that is on or off rather than one of a set. */
@Composable
fun BlueprintCheckRow(
    label: String,
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit,
    modifier: Modifier = Modifier,
) {
    OptionRow(
        label = label,
        modifier = modifier.toggleable(value = checked, role = Role.Checkbox, onValueChange = onCheckedChange),
        mark = { SelectionMark(selected = checked, filled = true) },
    )
}

@Composable
private fun OptionRow(label: String, modifier: Modifier, mark: @Composable () -> Unit) {
    Row(
        modifier = modifier.fillMaxWidth().defaultMinSize(minHeight = MinTouch),
        horizontalArrangement = Arrangement.spacedBy(Space.s),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        mark()
        Text(label, style = MaterialTheme.typography.bodyMedium, color = blueprint.text)
    }
}

/**
 * docs/09 "형태": no circles and no pills, so both marks are squares on the badge radius. A radio
 * holds a smaller square inside its outline; a checkbox fills, because "on" is a state and "this
 * one of the two" is a position.
 *
 * The outline is [app.recly.android.ui.theme.BlueprintColors.line] — the same token the hairline
 * and every bordered node take. An unchecked box is nothing *but* its outline.
 */
@Composable
private fun SelectionMark(selected: Boolean, filled: Boolean) {
    val palette = blueprint
    val ink = selectionInk(palette, selected)
    val shape = RoundedCornerShape(Radius.badge)
    Box(
        modifier = Modifier
            .size(MARK)
            .border(palette.line, ink, shape)
            .background(if (selected && filled) ink else Color.Transparent, shape),
        contentAlignment = Alignment.Center,
    ) {
        if (selected && !filled) Box(Modifier.size(INNER).background(ink, RoundedCornerShape(1.dp)))
        if (selected && filled) Box(Modifier.size(INNER).background(palette.onAccent, RoundedCornerShape(1.dp)))
    }
}

private val MARK: Dp = 18.dp
private val INNER: Dp = 8.dp

/** docs/09: a chosen option is the accent; an unchosen one takes the quiet border colour. */
fun selectionInk(palette: BlueprintColors, selected: Boolean): Color =
    if (selected) palette.accent else palette.textMuted

/**
 * A link inside a dialog — the Drive permissions page, the consent guidance. Accent and underlined,
 * because a link that is only a colour is invisible to a colour-blind reader (docs/09 "모든 상태는
 * 색 + 텍스트"), and [MinTouch] tall because it is something you tap.
 */
@Composable
fun BlueprintDialogLink(label: String, onClick: () -> Unit, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier
            .defaultMinSize(minHeight = MinTouch)
            .clickable(role = Role.Button, onClick = onClick),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            label,
            style = MaterialTheme.typography.bodyMedium,
            color = blueprint.accent,
            textDecoration = TextDecoration.Underline,
        )
    }
}
